File: lib/database_stack.py
"""Database stack for Aurora Global Database."""

from constructs import Construct
from cdktf_cdktf_provider_aws.rds_global_cluster import RdsGlobalCluster
from cdktf_cdktf_provider_aws.rds_cluster import RdsCluster
from cdktf_cdktf_provider_aws.rds_cluster_instance import RdsClusterInstance
from cdktf_cdktf_provider_aws.db_subnet_group import DbSubnetGroup
from cdktf_cdktf_provider_aws.rds_cluster_parameter_group import RdsClusterParameterGroup, RdsClusterParameterGroupParameter


class DatabaseStack(Construct):
    """Aurora Global Database infrastructure."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str,
        primary_provider,
        secondary_provider,
        primary_vpc_id: str,
        secondary_vpc_id: str,
        primary_subnet_ids: list,
        secondary_subnet_ids: list,
        primary_security_group_id: str,
        secondary_security_group_id: str,
    ):
        """Initialize Aurora Global Database in both regions."""
        super().__init__(scope, construct_id)

        # Create Aurora Global Cluster
        global_cluster = RdsGlobalCluster(
            self,
            "global_cluster",
            provider=primary_provider,
            global_cluster_identifier=f"aurora-global-{environment_suffix}",
            engine="aurora-mysql",
            engine_version="8.0.mysql_aurora.3.04.0",
            database_name="transactions",
            deletion_protection=False,  # Set to True for production
        )

        # Create DB subnet group in primary region
        primary_subnet_group = DbSubnetGroup(
            self,
            "primary_subnet_group",
            provider=primary_provider,
            name=f"primary-aurora-subnet-{environment_suffix}",
            subnet_ids=primary_subnet_ids,
            tags={
                "Name": f"primary-aurora-subnet-{environment_suffix}",
            },
        )

        # Create DB subnet group in secondary region
        secondary_subnet_group = DbSubnetGroup(
            self,
            "secondary_subnet_group",
            provider=secondary_provider,
            name=f"secondary-aurora-subnet-{environment_suffix}",
            subnet_ids=secondary_subnet_ids,
            tags={
                "Name": f"secondary-aurora-subnet-{environment_suffix}",
            },
        )

        # Create cluster parameter group for primary
        primary_param_group = RdsClusterParameterGroup(
            self,
            "primary_param_group",
            provider=primary_provider,
            name=f"primary-aurora-params-{environment_suffix}",
            family="aurora-mysql8.0",
            description="Parameter group for primary Aurora cluster",
            parameter=[
                RdsClusterParameterGroupParameter(
                    name="binlog_format",
                    value="ROW",
                    apply_method="pending-reboot"
                ),
                RdsClusterParameterGroupParameter(
                    name="character_set_server",
                    value="utf8mb4",
                    apply_method="pending-reboot"
                ),
            ],
            tags={
                "Name": f"primary-aurora-params-{environment_suffix}",
            },
        )

        # Create primary Aurora cluster in us-east-1
        primary_cluster = RdsCluster(
            self,
            "primary_cluster",
            provider=primary_provider,
            cluster_identifier=f"aurora-primary-{environment_suffix}",
            engine="aurora-mysql",
            engine_version="8.0.mysql_aurora.3.04.0",
            engine_mode="provisioned",
            database_name="transactions",
            master_username="admin",
            master_password="ChangeMe123!",  # Use AWS Secrets Manager in production
            db_subnet_group_name=primary_subnet_group.name,
            vpc_security_group_ids=[primary_security_group_id],
            db_cluster_parameter_group_name=primary_param_group.name,
            global_cluster_identifier=global_cluster.id,
            backup_retention_period=7,
            preferred_backup_window="03:00-04:00",
            preferred_maintenance_window="mon:04:00-mon:05:00",
            enabled_cloudwatch_logs_exports=["audit", "error", "general", "slowquery"],
            deletion_protection=False,  # Set to True for production
            skip_final_snapshot=True,  # Set to False for production
            serverlessv2_scaling_configuration={
                "min_capacity": 0.5,
                "max_capacity": 1.0,
            },
            depends_on=[global_cluster],
            tags={
                "Name": f"aurora-primary-{environment_suffix}",
                "Region": "us-east-1",
            },
        )

        # Create primary cluster instance (Serverless v2)
        primary_instance = RdsClusterInstance(
            self,
            "primary_instance",
            provider=primary_provider,
            identifier=f"aurora-primary-instance-1-{environment_suffix}",
            cluster_identifier=primary_cluster.id,
            instance_class="db.serverless",
            engine="aurora-mysql",
            engine_version="8.0.mysql_aurora.3.04.0",
            publicly_accessible=False,
            tags={
                "Name": f"aurora-primary-instance-1-{environment_suffix}",
            },
        )

        # Create secondary Aurora cluster in us-west-2
        secondary_cluster = RdsCluster(
            self,
            "secondary_cluster",
            provider=secondary_provider,
            cluster_identifier=f"aurora-secondary-{environment_suffix}",
            engine="aurora-mysql",
            engine_version="8.0.mysql_aurora.3.04.0",
            engine_mode="provisioned",
            db_subnet_group_name=secondary_subnet_group.name,
            vpc_security_group_ids=[secondary_security_group_id],
            global_cluster_identifier=global_cluster.id,
            backup_retention_period=7,
            preferred_backup_window="03:00-04:00",
            preferred_maintenance_window="mon:04:00-mon:05:00",
            enabled_cloudwatch_logs_exports=["audit", "error", "general", "slowquery"],
            deletion_protection=False,  # Set to True for production
            skip_final_snapshot=True,  # Set to False for production
            serverlessv2_scaling_configuration={
                "min_capacity": 0.5,
                "max_capacity": 1.0,
            },
            depends_on=[primary_cluster],
            tags={
                "Name": f"aurora-secondary-{environment_suffix}",
                "Region": "us-west-2",
            },
            lifecycle={
                "ignore_changes": ["master_username", "master_password"],
            },
        )

        # Create secondary cluster instance (Serverless v2)
        secondary_instance = RdsClusterInstance(
            self,
            "secondary_instance",
            provider=secondary_provider,
            identifier=f"aurora-secondary-instance-1-{environment_suffix}",
            cluster_identifier=secondary_cluster.id,
            instance_class="db.serverless",
            engine="aurora-mysql",
            engine_version="8.0.mysql_aurora.3.04.0",
            publicly_accessible=False,
            tags={
                "Name": f"aurora-secondary-instance-1-{environment_suffix}",
            },
        )

        # Export attributes for use in other stacks
        self.global_cluster_id = global_cluster.id
        self.primary_cluster_id = primary_cluster.id
        self.secondary_cluster_id = secondary_cluster.id
        self.primary_cluster_endpoint = primary_cluster.endpoint
        self.secondary_cluster_endpoint = secondary_cluster.endpoint
        self.primary_cluster_arn = primary_cluster.arn
        self.secondary_cluster_arn = secondary_cluster.arn

File: lib/database-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface DatabaseStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
  vpcId: pulumi.Input<string>;
  privateSubnetIds: pulumi.Input<string>[];
  kmsKeyId: pulumi.Input<string>;
}

export class DatabaseStack extends pulumi.ComponentResource {
  public readonly clusterEndpoint: pulumi.Output<string>;
  public readonly clusterArn: pulumi.Output<string>;

  constructor(
    name: string,
    args: DatabaseStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:database:DatabaseStack', name, args, opts);

    const { environmentSuffix, tags, vpcId, privateSubnetIds, kmsKeyId } = args;

    // For production, use AWS Secrets Manager to store the password
    // For testing/demo purposes, we use a simple password
    // NOTE: This should be replaced with proper secret management in production
    const dbPassword = `TempPassword${environmentSuffix}123!`;

    // Security group for RDS
    const dbSecurityGroup = new aws.ec2.SecurityGroup(
      `payment-db-sg-${environmentSuffix}`,
      {
        vpcId: vpcId,
        description: 'Security group for Aurora PostgreSQL cluster',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 5432,
            toPort: 5432,
            cidrBlocks: ['10.0.0.0/16'],
            description: 'PostgreSQL from VPC',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound',
          },
        ],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-db-sg-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // DB subnet group
    const dbSubnetGroup = new aws.rds.SubnetGroup(
      `payment-db-subnet-group-${environmentSuffix}`,
      {
        subnetIds: privateSubnetIds,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-db-subnet-group-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // DB parameter group for PostgreSQL
    const dbParameterGroup = new aws.rds.ClusterParameterGroup(
      `payment-db-params-${environmentSuffix}`,
      {
        family: 'aurora-postgresql14',
        description: 'Parameter group for payment processing database',
        parameters: [
          {
            name: 'log_statement',
            value: 'all',
          },
          {
            name: 'log_min_duration_statement',
            value: '1000',
          },
          {
            name: 'ssl',
            value: '1',
          },
        ],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-db-params-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Aurora PostgreSQL cluster
    const cluster = new aws.rds.Cluster(
      `payment-cluster-${environmentSuffix}`,
      {
        clusterIdentifier: `payment-cluster-${environmentSuffix}`,
        engine: 'aurora-postgresql',
        engineVersion: '14.6',
        databaseName: 'paymentdb',
        masterUsername: 'dbadmin',
        masterPassword: dbPassword,
        dbSubnetGroupName: dbSubnetGroup.name,
        vpcSecurityGroupIds: [dbSecurityGroup.id],
        dbClusterParameterGroupName: dbParameterGroup.name,
        storageEncrypted: true,
        kmsKeyId: kmsKeyId,
        backupRetentionPeriod: 30,
        preferredBackupWindow: '03:00-04:00',
        preferredMaintenanceWindow: 'sun:04:00-sun:05:00',
        enabledCloudwatchLogsExports: ['postgresql'],
        skipFinalSnapshot: true,
        copyTagsToSnapshot: true,
        deletionProtection: false,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-cluster-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Primary instance
    const primaryInstance = new aws.rds.ClusterInstance(
      `payment-instance-1-${environmentSuffix}`,
      {
        identifier: `payment-instance-1-${environmentSuffix}`,
        clusterIdentifier: cluster.id,
        instanceClass: 'db.t3.medium',
        engine: 'aurora-postgresql',
        engineVersion: '14.6',
        publiclyAccessible: false,
        performanceInsightsEnabled: true,
        performanceInsightsKmsKeyId: kmsKeyId,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-instance-1-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Reader instance 1
    const readerInstance1 = new aws.rds.ClusterInstance(
      `payment-instance-2-${environmentSuffix}`,
      {
        identifier: `payment-instance-2-${environmentSuffix}`,
        clusterIdentifier: cluster.id,
        instanceClass: 'db.t3.medium',
        engine: 'aurora-postgresql',
        engineVersion: '14.6',
        publiclyAccessible: false,
        performanceInsightsEnabled: true,
        performanceInsightsKmsKeyId: kmsKeyId,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-instance-2-${environmentSuffix}`,
        })),
      },
      { parent: this, dependsOn: [primaryInstance] }
    );

    // Reader instance 2
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const readerInstance2 = new aws.rds.ClusterInstance(
      `payment-instance-3-${environmentSuffix}`,
      {
        identifier: `payment-instance-3-${environmentSuffix}`,
        clusterIdentifier: cluster.id,
        instanceClass: 'db.t3.medium',
        engine: 'aurora-postgresql',
        engineVersion: '14.6',
        publiclyAccessible: false,
        performanceInsightsEnabled: true,
        performanceInsightsKmsKeyId: kmsKeyId,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-instance-3-${environmentSuffix}`,
        })),
      },
      { parent: this, dependsOn: [readerInstance1] }
    );

    this.clusterEndpoint = cluster.endpoint;
    this.clusterArn = cluster.arn;

    this.registerOutputs({
      clusterEndpoint: this.clusterEndpoint,
      clusterArn: this.clusterArn,
      readerEndpoint: cluster.readerEndpoint,
    });
  }
}
